### PR TITLE
Change our code review guidelines (for this repository)

### PR DESCRIPTION
**Pull request by yuvipanda**
This incorporates the de-facto guidelines we have been operating
under for a few weeks now. I've received 1:1 input from most of
the folks who work on this repository, and incorporated that into
this policy. It also derives a lot of its foundation from
https://github.com/2i2c-org/infrastructure/issues/3997#issuecomment-2079956224.

At this point, I think it's safe for us to try for a few weeks
and see how this goes.

I've split out the guidelines for community partners (which have not
changed at all) into its own page. I've also removed extra, specific
guidelines for auth changes and terraform - I think the broad overall
guidelines cover these as well now.

Fixes https://github.com/2i2c-org/infrastructure/issues/3997

### CODE REVIEW

**Review by consideRatio — APPROVED** [review]
Looks great, I greatly appreciate this change and effort @yuvipanda!

Since the documented changes has been knowledge shared sync over time, I think it makes sense for me to hit merge here rather than to request/wait for additional people in engineering to review this document.

File: docs/contributing/code-review.md [hunk 1]
```diff
@@ -1,235 +1,118 @@
...
-   organizational restrictions can cause it to fail. Timeouts can happen.
-   Runtime errors pop up.
+You should break this rule when it makes sense, but do not do so
+**casually**. This is an important mechanism by which we can build a
+team we can all trust in over
 
-Thanks to all these, it is impossible to either:
+This may make you uncomfortable, and that is ok. Ultimately our goal is to
+build a **team we can all trust in**, and this is a very important part
+of getting there.
```
Comment (sgibson91) [inline docs/contributing/code-review.md:69]: This is maybe a repeat of the above unfinished sentence?

File: docs/contributing/code-review.md [hunk 2]
```diff
@@ -1,235 +1,118 @@
...
-As a result, we have some workflows that are specific to using Terraform, described below.
+After you make your PR, ask yourself these three questions:
 
-### Changes to running infrastructure
+1. *If* this PR breaks something, do you feel comfortable debugging it **or**
+   reverting it to undo the change?
+2. Do you know how to *verify* the effects of merging this PR?
 
-In this scenario, there is already-running infrastructure and you are deploying changes to it.
+If the answer to all these 2 questions are 'yes', you should go ahead
```
Comment (sgibson91) [inline docs/contributing/code-review.md:79]: Again, I would exclude the "2" here so we can add to the list of questions without needing to update other language.

**Comment by yuvipanda** [comment]
Thank you @sgibson91! I've opened (and merged!) https://github.com/2i2c-org/infrastructure/pull/4322 to address your feedback.

Also love the attitude and the merge, @consideRatio! Thank you.

**Pull request merged**